test/integration/apiserver/flowcontrol/concurrency_util_test.go

```diff
@@ -0,0 +1,496 @@
+	t.Logf("\nnoxu1PLSeatUtilAvg %v\nnoxu2PLSeatUtilAvg %v", noxu1PLSeatUtilAvg, noxu2PLSeatUtilAvg)
+
+	wg.Wait() // wait till the client goroutines finish before computing the statistics
+	noxu1ClientRequestLatencySecondsAvg, noxu1ClientRequestLatencySecondsSdev := computeClientRequestLatencyStats(noxu1ClientRequestLatencyCount, noxu1ClientRequestLatencySum, noxu1ClientRequestLatencySumSq)
+	noxu2ClientRequestLatencySecondsAvg, noxu2ClientRequestLatencySecondsSdev := computeClientRequestLatencyStats(noxu2ClientRequestLatencyCount, noxu2ClientRequestLatencySum, noxu2ClientRequestLatencySumSq)
+	t.Logf("\nnoxu1ClientRequestLatencyCount %v\nnoxu2ClientRequestLatencyCount %v", noxu1ClientRequestLatencyCount, noxu2ClientRequestLatencyCount)
+	t.Logf("\nnoxu1ClientRequestLatencySecondsAvg %v\nnoxu2ClientRequestLatencySecondsAvg %v", noxu1ClientRequestLatencySecondsAvg, noxu2ClientRequestLatencySecondsAvg)
+	t.Logf("\nnoxu1ClientRequestLatencySecondsSdev %v\nnoxu2ClientRequestLatencySecondsSdev %v", noxu1ClientRequestLatencySecondsSdev, noxu2ClientRequestLatencySecondsSdev)
+	allDispatchedReqCounts, rejectedReqCounts, err := getRequestCountOfPriorityLevel(loopbackClient)
+	if err != nil {
+		t.Error(err)
+	}
+	t.Logf("\nnoxu1APFRequestCount %v\nnoxu2APFRequestCount %v", allDispatchedReqCounts[priorityLevelNoxu1.Name], allDispatchedReqCounts[priorityLevelNoxu2.Name])
+	if rejectedReqCounts[priorityLevelNoxu1.Name] > 0 {
+		t.Errorf(`%v requests from the "elephant" stream were rejected unexpectedly`, rejectedReqCounts[priorityLevelNoxu2.Name])
+	}
+	if rejectedReqCounts[priorityLevelNoxu2.Name] > 0 {
+		t.Errorf(`%v requests from the "mouse" stream were rejected unexpectedly`, rejectedReqCounts[priorityLevelNoxu2.Name])
+	}
+
+	// Calculate server-side observed concurrency
+	noxu1ObservedConcurrency := noxu1PLSeatUtilAvg * float64(availableSeats[priorityLevelNoxu1.Name])
+	noxu2ObservedConcurrency := noxu2PLSeatUtilAvg * float64(availableSeats[priorityLevelNoxu2.Name])
+	// Expected concurrency is derived from equal throughput assumption on both the client-side and the server-side
+	// Expected concurrency computed can sometimes be larger than the number of available seats. We use the number of available seats as an upper bound
+	noxu1ExpectedConcurrency := math.Min(float64(noxu1NumGoroutines)*noxu1RequestExecutionSecondsAvg/noxu1ClientRequestLatencySecondsAvg, float64(availableSeats[priorityLevelNoxu1.Name]))
+	noxu2ExpectedConcurrency := math.Min(float64(noxu2NumGoroutines)*noxu2RequestExecutionSecondsAvg/noxu2ClientRequestLatencySecondsAvg, float64(availableSeats[priorityLevelNoxu2.Name]))
+	t.Logf("Concurrency of noxu1:noxu2 - expected (%v:%v), observed (%v:%v)", noxu1ExpectedConcurrency, noxu2ExpectedConcurrency, noxu1ObservedConcurrency, noxu2ObservedConcurrency)
+	// Calculate the tolerable error margin and perform the final check
```

Comment:
Please give a reference to the statistical test that we're doing here, or otherwise explain it?

Reply:
I am concerned that this is going to be too flaky, how many times have you run this test? Flakes definitely need to be < 1/100, ideally < 1/1000.

Reply:
This will be relevant: https://en.wikipedia.org/wiki/Error_function

Reply:
I would have expected a call to https://pkg.go.dev/math#Erf then, somewhere?

Reply:
@lavalamp could you suggest some way to show the failure rate is low (i.e. not flaky) in CI? I can test in my environment thousands of times more, but I don't think it's meaningful if my environment is not as noisy as in CI

Reply:
https://github.com/kubernetes/kubernetes/pull/108553#issuecomment-1064871247

compile the test and run it with `stress -test.run Test..`

Reply:
@aojea thanks for the tip. This revealed some failures from the checks of hard-coded error bound numbers. Will try to adjust them

Reply:
Thanks. You can perhaps add a bit more delay to the webhook to increase the signal, if you need to?

Also it's looking much more readable after the last push, thanks.

Reply:
To answer the original question, the idea is very basic.  We know there are uncontrolled overheads that introduce noise into the system.  The coefficient of variation --- that is, standard deviation divided by mean --- for a class of traffic is a characterization of all the noise that applied to that class.  We found that class 1 generally had a much bigger CV than class 2.  This makes sense, because class 1 probes more behavior --- the waiting in queues.  So we take the minimum of the two as an indicator of the relative amount of noise that comes from all the other behavior.

We picked two standard deviations because it is commonly used as a discriminator between common and uncommon.  We do not need to call the error function at runtime, because we are using a fixed multiplier of the standard deviation.  Using 2 as that multiplier gets the result that we expect about 95.4% of trials to fall within the bound.  A multiplier of 3 would expect 99.7% to fall within the bound.

Of course, the above is assuming that the noise is normally distributed, which we are not really justified in assuming.  So all this is really random precision.  We just wanted some way to make the test adjust to the noise it experienced.

Reply:
I'll incorporate Mike's comment into the code comment as the explanation on the margin of error calculation

Reply:
I stress tested the latest version
```
$ stress ./flowcontrol.test -test.run=TestConcurrencyIsolation
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 0 runs so far, 0 failures
20s: 0 runs so far, 0 failures
25s: 0 runs so far, 0 failures
30s: 0 runs so far, 0 failures
35s: 16 runs so far, 0 failures
40s: 16 runs so far, 0 failures
45s: 16 runs so far, 0 failures
50s: 16 runs so far, 0 failures
55s: 16 runs so far, 0 failures
1m0s: 18 runs so far, 0 failures
...
33m55s: 1059 runs so far, 0 failures
34m0s: 1061 runs so far, 0 failures
34m5s: 1064 runs so far, 0 failures
34m10s: 1066 runs so far, 0 failures
34m15s: 1069 runs so far, 0 failures
```